xapian-core/common/serialise-double.cc

```diff
@@ -38,175 +40,152 @@
+    result |= scaled_v_int;
+
+# ifdef WORDS_BIGENDIAN
+    result = do_bswap(result);
+# endif
+
+    return string(reinterpret_cast<const char *>(&result), sizeof(uint64_t));
 }
 
-double unserialise_double(const char ** p, const char *end)
-{
-    if (end - *p < 2) {
-	throw Xapian::SerialisationError("Bad encoded double: insufficient data");
-    }
-    unsigned char first = *(*p)++;
-    if (first == 0 && *(*p) == 0) {
-	++*p;
-	return 0.0;
+double unserialise_double(const char ** p, const char * end) {
+    if (end - *p < 8) {
+	throw Xapian::SerialisationError(
+	    "Bad encoded double: insufficient data");
     }
+    unsigned char first = *(*p + 7); // little-endian stored
+    unsigned char second = *(*p + 6);
 
-    bool negative = (first & 0x80) != 0;
-    size_t mantissa_len = ((first >> 4) & 0x07) + 1;
-
-    int exp = first & 0x0f;
-    if (exp >= 14) {
-	int bigexp = static_cast<unsigned char>(*(*p)++);
-	if (exp == 15) {
-	    if (*p == end) {
-		throw Xapian::SerialisationError("Bad encoded double: short large exponent");
-	    }
-	    exp = bigexp | (static_cast<unsigned char>(*(*p)++) << 8);
-	    exp -= 32768;
-	} else {
-	    exp = bigexp - 128;
-	}
-    } else {
-	exp -= 7;
-    }
+    bool negative = (first & (0x80)) != 0;
 
-    if (size_t(end - *p) < mantissa_len) {
-	throw Xapian::SerialisationError("Bad encoded double: short mantissa");
-    }
+    // bitwise operations to extract exponent
+    int exp = (first & (0x80 - 1));
+    exp <<= 4;
+    exp |= (second & (15 << 4)) >> 4;
+    exp -= 1023;
 
-    double v = 0.0;
-
-    static double dbl_max_mantissa = DBL_MAX;
-    static int dbl_max_exp = base256ify_double(dbl_max_mantissa);
-    *p += mantissa_len;
-    if (exp > dbl_max_exp ||
-	(exp == dbl_max_exp &&
-	 double(static_cast<unsigned char>((*p)[-1])) > dbl_max_mantissa)) {
-	// The mantissa check should be precise provided that FLT_RADIX
-	// is a power of 2.
-	v = HUGE_VAL;
-    } else {
-	const char *q = *p;
-	while (mantissa_len--) {
-	    v *= 0.00390625; // 1/256
-	    v += double(static_cast<unsigned char>(*--q));
-	}
+    double mantissa = 1.0; // implicit 1 in IEEE
+    double pows2 = 0.5; // powers of 2
 
-#if FLT_RADIX == 2
-	if (exp) v = scalbn(v, exp * 8);
-#elif FLT_RADIX == 16
-	if (exp) v = scalbn(v, exp * 2);
-#else
-	if (exp) v = ldexp(v, exp * 8);
-#endif
+    int second_byte_counter = 1 << 3;
+    for (int i = 0; i < 4; ++i) {
+	if (second & second_byte_counter) mantissa += pows2;
+	pows2 /= 2.0;
+	second_byte_counter >>= 1;
+    }
 
-#if 0
-	if (v == 0.0) {
-	    // FIXME: handle underflow
+    int byte_counter = 5; // starting with the third byte onwards
+    while (byte_counter >= 0) {
+	unsigned char cur_byte = *(*p + byte_counter);
+	--byte_counter;
+	int cur_byte_counter = 1 << 7;
+	for (int i = 0; i < 8; ++i) {
+	    if (cur_byte & cur_byte_counter) mantissa += pows2;
+	    pows2 /= 2.0;
+	    cur_byte_counter >>= 1;
 	}
-#endif
```

Comment:
Again this is processing one bit at a time, which seems complex and inefficient.

Can we unserialise by doing the reverse the steps we use to serialise?

Reply:
Fixed it in the latest commit.